sdks/java/harness/src/main/java/org/apache/beam/fn/harness/status/BeamFnStatusClient.java

```diff
@@ -0,0 +1,120 @@
+    BeamFnWorkerStatusGrpc.BeamFnWorkerStatusStub stub =
+        BeamFnWorkerStatusGrpc.newStub(channelFactory.apply(apiServiceDescriptor));
+    stub.workerStatus(new InboundObserver());
+    this.outboundObserver = stub.workerStatus(new InboundObserver());
+    this.processBundleCache = processBundleCache;
+  }
+
+  void threadDump(StringJoiner trace) {
+    trace.add("========== THREAD DUMP ==========");
+    Map<Thread, StackTraceElement[]> threads = Thread.getAllStackTraces();
+    threads
+        .entrySet()
+        .forEach(
+            entry -> {
+              trace.add(
+                  String.format(
+                      "%n--- Thread #%d name: %s ---",
+                      entry.getKey().getId(), entry.getKey().getName()));
+              Arrays.stream(entry.getValue()).map(StackTraceElement::toString).forEach(trace::add);
+            });
+  }
+
+  @VisibleForTesting
+  void activeProcessBundleState(StringJoiner activeBundlesState) {
+    activeBundlesState.add("========== ACTIVE PROCESSING BUNDLES ==========");
+    if (processBundleCache.getActiveBundleProcessors().isEmpty()) {
+      activeBundlesState.add("No active processing bundles.\n");
+    } else {
+      processBundleCache.getActiveBundleProcessors().entrySet().stream()
+          .sorted(
+              Comparator.comparingLong(
+                      (Map.Entry<String, BundleProcessor> bundle) ->
+                          bundle.getValue().getStateTracker().getMillisSinceLastTransition())
+                  .reversed()) // reverse sort active bundle by time since last transition.
+          .limit(10) // only keep top 10
+          .forEach(
+              entry -> {
+                ExecutionStateTracker executionStateTracker = entry.getValue().getStateTracker();
+                activeBundlesState.add(String.format("---- Instruction %s ----", entry.getKey()));
+                activeBundlesState.add(
+                    String.format(
+                        "Tracked thread: %s", executionStateTracker.getTrackedThread().getName()));
+                activeBundlesState.add(
+                    String.format(
+                        "Time since transition: %.2f seconds%n",
+                        executionStateTracker.getMillisSinceLastTransition() / 1000.0));
+              });
+    }
+  }
+
+  private class InboundObserver implements StreamObserver<BeamFnApi.WorkerStatusRequest> {
+    @Override
+    public void onNext(WorkerStatusRequest workerStatusRequest) {
+      StringJoiner status = new StringJoiner("\n");
+      activeProcessBundleState(status);
+      status.add("\n");
+      threadDump(status);
+      outboundObserver.onNext(
+          WorkerStatusResponse.newBuilder()
+              .setId(workerStatusRequest.getId())
+              .setStatusInfo(status.toString())
+              .build());
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      LOG.error("Error getting SDK harness status", t);
```

Comment:
We can also sends the error back to be captured.

Reply:
I think outboundObserver.onError(t) will not actually send back the exception as an error response but instead shut down the grpc connection. We can consider either just logging the error without calling outboundObserver.onError(t), or try to catch exception in onNext() and send it back with 
```
WorkerStatusResponse.newBuilder()
              .setId(workerStatusRequest.getId())
              .setError(exceptionMesasge)
              .build()
```